### PR TITLE
FEAT: Streaming support in fetchone, fetchmany, fetchall for varbinarymax data type.

### PR DESCRIPTION
**Pull request by gargsaumya**
### Work Item / Issue Reference  

> [AB#33395](https://sqlclientdrivers.visualstudio.com/c6d89619-62de-46a0-8b46-70b92a84d85e/_workitems/edit/33395)

> GitHub Issue: #<ISSUE_NUMBER>

-------------------------------------------------------------------
### Summary   
  
This pull request significantly improves support for streaming and fetching large binary (VARBINARY(MAX)) and large text columns in the MSSQL Python driver. The main changes include robust chunked retrieval of large objects (LOBs), correct handling of edge cases (such as empty or null values), and enhanced test coverage for these scenarios.

**LOB Streaming and Fetching Enhancements:**

* Added a new helper function `FetchLobColumnData` to efficiently stream and assemble large binary/text columns (LOBs) from the database, handling chunking, null/empty values, and correct type conversion for both binary and (wide/narrow) string columns.
* Updated the logic in `SQLGetData_wrap`, `FetchBatchData`, `FetchMany_wrap`, and `FetchAll_wrap` to detect LOB columns and use the new streaming path for fetching them, including proper fallback to row-by-row fetching when LOBs are present. [[1]](diffhunk://#diff-dde2297345718ec449a14e7dff91b7bb2342b008ecc071f562233646d71144a1L2062-R2184) [[2]](diffhunk://#diff-dde2297345718ec449a14e7dff91b7bb2342b008ecc071f562233646d71144a1L2542-R2631) [[3]](diffhunk://#diff-dde2297345718ec449a14e7dff91b7bb2342b008ecc071f562233646d71144a1R2760-R2788) [[4]](diffhunk://#diff-dde2297345718ec449a14e7dff91b7bb2342b008ecc071f562233646d71144a1L2699-R2803) [[5]](diffhunk://#diff-dde2297345718ec449a14e7dff91b7bb2342b008ecc071f562233646d71144a1R2882-R2910) [[6]](diffhunk://#diff-dde2297345718ec449a14e7dff91b7bb2342b008ecc071f562233646d71144a1L2792-R2925)
* Modified the batch fetch function signatures and logic to propagate LOB column information and ensure correct handling during bulk fetches.

**Testing Improvements:**

* Replaced and expanded the test for large binary data with a new, comprehensive test (`test_varbinarymax_insert_fetch`) that verifies insertion and retrieval of empty, small, and large VARBINARY(MAX) values (including edge cases around the 8000-byte threshold) using `fetchone`, `fetchall`, and `fetchmany`.

These changes ensure that the driver can reliably handle large binary and text columns in all fetch scenarios, improving correctness and robustness for users working with LOB data.

### CODE REVIEW

File: mssql_python/pybind/ddbc_bindings.cpp [hunk 1]
```diff
@@ -1751,6 +1751,96 @@ SQLRETURN SQLFetch_wrap(SqlHandlePtr StatementHandle) {
...
+            cType,
+            chunk.data(),
+            DAE_CHUNK_SIZE,
+            &indicator
+        );
+        if (indicator == SQL_NULL_DATA) {
+            LOG("Loop {}: Column {} is NULL", loopCount, colIndex);
+            return py::none();
+        }
+        if (!SQL_SUCCEEDED(ret) && ret != SQL_SUCCESS_WITH_INFO) {
```
Comment (bewithgaurav) [inline mssql_python/pybind/ddbc_bindings.cpp]: can we put this block on line 1776? just after `SQLGetData_ptr`?

File: mssql_python/pybind/ddbc_bindings.cpp [hunk 2]
```diff
@@ -2059,45 +2149,39 @@ SQLRETURN SQLGetData_wrap(SqlHandlePtr StatementHandle, SQLUSMALLINT colCount, p
...
+                                oss << "Buffer length for fetch (" << columnSize << ") is smaller than actual data ("
+                                    << dataLen << "). ColumnID - " << i << ", datatype - " << dataType;
+                                ThrowStdException(oss.str());
+                            }
+                        } else if (dataLen == SQL_NULL_DATA) {
+                            row.append(py::none());
+                        } else if (dataLen == 0) {
+                            row.append(py::bytes(""));
+                        } else {
+                            LOG("SQLGetData returned unexpected negative length: {}. Column ID - {}", dataLen, i);
```
Comment (bewithgaurav) [inline mssql_python/pybind/ddbc_bindings.cpp]: could we also add dataType in LOG? would be better to keep this LOG consistent across dataTypes
below is snippet from `SQL_LONGVARCHAR`
```
// This is unexpected
                        LOG("SQLGetData returned an unexpected negative data length. "
                            "Raising exception. Column ID - {}, Data Type - {}, Data Length - {}",
                            i, dataType, dataLen);
                        ThrowStdException("SQLGetData returned an unexpected negative data length");
 ```

File: tests/test_004_cursor.py [hunk 3]
```diff
@@ -6309,34 +6309,71 @@ def test_binary_data_over_8000_bytes(cursor, db_connection):
...
+        # ---------- FETCHMANY TEST ----------
+        cursor.execute("SELECT id, binary_data FROM #pytest_varbinarymax ORDER BY id")
+        batch_size = 2
+        batches = []
+        while True:
+            batch = cursor.fetchmany(batch_size)
+            if not batch:
+                break
+            batches.extend(batch)
+        assert len(batches) == len(test_data)
```
Comment (bewithgaurav) [inline tests/test_004_cursor.py:6175]: nice testcase, tests all fetch methods and datatype edge cases. can we also test streaming using executemany?
add exactly same testcase with executemany as well if possible - will cover everything

File: mssql_python/pybind/ddbc_bindings.cpp (hunk 2 above)
Comment (gargsaumya) [reply mssql_python/pybind/ddbc_bindings.cpp]: Yes this is already handled in previous PRs. This PR will get rebased once the previous PR merges.

File: mssql_python/pybind/ddbc_bindings.cpp (hunk 1 above)
Comment (gargsaumya) [reply mssql_python/pybind/ddbc_bindings.cpp]: addressed below.

File: tests/test_004_cursor.py (hunk 3 above)
Comment (gargsaumya) [reply tests/test_004_cursor.py:6175]: streaming in executemany will be a separate work item and will be covered in future PR. Noted.

File: mssql_python/pybind/ddbc_bindings.cpp [hunk 4]
```diff
@@ -2155,45 +2155,39 @@ SQLRETURN SQLGetData_wrap(SqlHandlePtr StatementHandle, SQLUSMALLINT colCount, p
...
+                            ThrowStdException("Unexpected negative SQLGetData length");
```
Comment (sumitmsft) [inline mssql_python/pybind/ddbc_bindings.cpp]: can we please provide more detailed messages here? Like information such as the column index, data type, and relevant variable values (e.g., buffer size, returned length). This will help in debugging

File: mssql_python/pybind/ddbc_bindings.cpp [hunk 5]
```diff
@@ -2155,45 +2155,39 @@ SQLRETURN SQLGetData_wrap(SqlHandlePtr StatementHandle, SQLUSMALLINT colCount, p
...
-						} else {
-                            // In this case, buffer size is smaller, and data to be retrieved is longer
-                            // TODO: Revisit
-                            std::ostringstream oss;
-                            oss << "Buffer length for fetch (" << columnSize << ") is smaller, & data "
-                                << "to be retrieved is longer (" << dataLen << "). ColumnID - "
-                                << i << ", datatype - " << dataType;
-                            ThrowStdException(oss.str());
+                    if (SQL_SUCCEEDED(ret)) {
+                        if (dataLen > 0) {
```
Comment (sumitmsft) [inline mssql_python/pybind/ddbc_bindings.cpp:2169]: May be for future enhancement... 

We have code duplication in how data length checks and error handling are performed when fetching data in `SQLGetData_wrap` for multiple data types. 
Can we refactoring these patterns into a helper function to reduce the risk of future bugs?

File: tests/test_004_cursor.py [hunk 6]
```diff
@@ -6113,34 +6113,71 @@ def test_binary_data_over_8000_bytes(cursor, db_connection):
...
+
```
Comment (sumitmsft) [inline tests/test_004_cursor.py:6176]: Can we consider adding a test for NULL binary values to ensure they are handled as expected?

**Review by sumitmsft — CHANGES_REQUESTED** [review]
Minor fixes are required.

**Review by sumitmsft — APPROVED** [review]
All looks good.

**Review by bewithgaurav — APPROVED** [review]
